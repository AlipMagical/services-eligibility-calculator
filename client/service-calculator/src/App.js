/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import './i18n';

import React, { createContext, useState } from 'react';

import { ANALYTICS_ID } from './common/config';
import { Body } from './components/Body';
import { DataContext } from './DataContext';
import Footer from './components/Footer';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import { Router } from 'react-router-dom';
import ScrollToTop from './common/utils/ScrollToTop';
import { createBrowserHistory } from 'history';

export const FilterContext = createContext();

export const createInitialFilterState = () => ({
  income: undefined,
  interests: {},
  people: {}
});

const history = createBrowserHistory();

export default function App(props) {
  const [analyticsInitialized, setAnalyticsInitialized] = useState(false);
  const [lastSearch, setLastSearch] = useState('');
  const [filterState, setFilterState] = useState(createInitialFilterState());
  const resetFilterState = () => setFilterState(createInitialFilterState());

  if (!analyticsInitialized) {
    ReactGA.initialize(ANALYTICS_ID, { testMode: props.testMode });
    setAnalyticsInitialized(true);
  }

  return (
    <DataContext.StateProvider>
      <FilterContext.Provider
        value={{
          filterState,
          setFilterState,
          resetFilterState,
          lastSearch,
          setLastSearch
        }}
      >
        <Router history={history}>
          <ScrollToTop />
          <Body />
          <Footer />
        </Router>
      </FilterContext.Provider>
    </DataContext.StateProvider>
  );
}

App.propTypes = {
  testMode: PropTypes.bool
};
