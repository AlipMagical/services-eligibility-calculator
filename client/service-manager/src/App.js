/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import './i18n';

import { BrowserRouter, Redirect, Route, Switch } from 'react-router-dom';
import React, { createContext, useState } from 'react';
import {
  getLoggedInUser,
  getMillisUntilExpiration,
  isLoggedIn
} from './common/auth';

import { ANALYTICS_ID } from './common/config';
import { BASEPATH } from './common/config';
import Header from './components/Header';
import PropTypes from 'prop-types';
import ReactGA from 'react-ga';
import Toast from './components/Toast';
import departmentRoutes from './routes/department';
import getUserConfirmation from './getUserConfirmation';
import interestRoutes from './routes/interest';
import serviceRoutes from './routes/service';
import settingsRoutes from './routes/settings';
import { useSettings } from './hooks';

export const AuthContext = createContext();
export const ToastContext = createContext();
export const SettingsContext = createContext();

export default function App(props) {
  ReactGA.initialize(ANALYTICS_ID, { testMode: props.testMode });

  const [loginState, setLoginState] = useState({
    isLoggedIn: isLoggedIn(),
    loggedInUser: getLoggedInUser()
  });

  const onLoginStatusChanged = () => {
    //setTimeout to update when token might expire.
    const millis = getMillisUntilExpiration();
    millis && setTimeout(onLoginStatusChanged, millis);

    setLoginState({
      isLoggedIn: isLoggedIn(),
      loggedInUser: getLoggedInUser()
    });
  };

  const { settings, setSettings } = useSettings();

  const [toastMessage, setToastMessage] = useState(null);

  return (
    <AuthContext.Provider value={{ ...loginState, onLoginStatusChanged }}>
      <SettingsContext.Provider value={{ settings, setSettings }}>
        <ToastContext.Provider value={{ toastMessage, setToastMessage }}>
          <BrowserRouter getUserConfirmation={getUserConfirmation}>
            <Header />
            <div className="container bd-content">
              <Switch>
                {serviceRoutes}
                {settingsRoutes}
                {interestRoutes}
                {departmentRoutes}
                <Redirect from={BASEPATH} exact to={BASEPATH + 'services/'} />
                <Route render={() => <h1>Not Found</h1>} />
              </Switch>
            </div>
            {toastMessage && <Toast message={toastMessage} />}
          </BrowserRouter>
        </ToastContext.Provider>
      </SettingsContext.Provider>
    </AuthContext.Provider>
  );
}

App.propTypes = {
  testMode: PropTypes.bool
};
