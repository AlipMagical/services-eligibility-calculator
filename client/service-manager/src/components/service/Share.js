/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Col, Row } from 'react-bootstrap';
import { Field, Form as FinalForm } from 'react-final-form';

import ActionModal from '../ActionModal';
import PropTypes from 'prop-types';
import { REQUIRED_DOMAIN_NAME } from '../../config';
import React from 'react';
import { ToastContext } from '../../App';
import { fetch } from '../../common/dataAccess';
import { sendEmail } from '../../common/utils';

export default function Share(props) {
  const { setToastMessage } = React.useContext(ToastContext);
  const isDraft = props.serviceUrl.includes('service_drafts');

  let onSubmit = async values => {
    var email = values.email;
    var username = email.substring(0, email.indexOf('@'));
    fetch(isDraft ? 'service_draft_shares' : 'service_shares', {
      method: 'POST',
      body: JSON.stringify({
        sharedWithUsername: username,
        service: props.serviceUrl,
        draftService: props.serviceUrl
      })
    })
      .then(() => {
        setToastMessage(props.serviceName + ' shared with ' + email);
        var emailBody =
          "I've shared the following service with you:%0D%0A%0D%0A";
        emailBody += props.serviceName + '%0D%0A';
        emailBody += encodeURIComponent(window.location.href);
        sendEmail(email, 'Shared Service', emailBody);
        props.onShared(username);
      })
      .catch(error => {
        window.alert('Failed to share service.\n' + error);
        props.onCancel();
      });
  };

  let validateEmail = value => {
    var emailFormat = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;
    var validationErrors = emailFormat.test(value)
      ? undefined
      : 'Invalid email address';
    if (
      !validationErrors &&
      REQUIRED_DOMAIN_NAME &&
      REQUIRED_DOMAIN_NAME.length > 0
    ) {
      var domainName = value.substring(value.indexOf('@') + 1, value.length);
      validationErrors =
        domainName === REQUIRED_DOMAIN_NAME
          ? undefined
          : 'Address must be  @' + REQUIRED_DOMAIN_NAME;
    }
    return validationErrors;
  };

  let renderField = data => {
    data.input.className = 'form-control';

    const isInvalid = data.meta.error && data.meta.touched;
    if (isInvalid) {
      data.input.className += ' is-invalid';
      data.input['aria-invalid'] = true;
    }

    return (
      <>
        <Row>
          <Col sm={3}>
            <label
              htmlFor={`share_${data.input.name}`}
              className={'col-form-label' + (isInvalid ? ' invalid' : '')}
            >
              {data.label}
            </label>
          </Col>
          <Col>
            <input
              {...data.input}
              required={data.required}
              id={`share_${data.input.name}`}
            />
          </Col>
        </Row>
        {data.meta.error && data.meta.touched && (
          <Row>
            <Col sm={3} />
            <Col className="error">{data.meta.error}</Col>
          </Row>
        )}
      </>
    );
  };

  return (
    <ActionModal
      title="Share service"
      analyticsLabel="Share"
      cancelText="Cancel"
      onCancel={props.onCancel}
      actionText="Share"
      onAction={() =>
        document.getElementById('share-form').dispatchEvent(new Event('submit'))
      }
      renderBody={footer => (
        <>
          <p>They will be able to edit, publish, and delete this service</p>
          <FinalForm
            onSubmit={onSubmit}
            render={({ handleSubmit }) => (
              <form id={'share-form'} onSubmit={handleSubmit}>
                <div className="form-body">
                  <Field
                    render={renderField}
                    label="Email:"
                    name="email"
                    type="text"
                    required={true}
                    validate={validateEmail}
                  />
                </div>
                {footer}
              </form>
            )}
          />
        </>
      )}
    />
  );
}

Share.propTypes = {
  serviceUrl: PropTypes.string.isRequired,
  serviceName: PropTypes.string.isRequired,
  onCancel: PropTypes.func,
  onShare: PropTypes.func
};
