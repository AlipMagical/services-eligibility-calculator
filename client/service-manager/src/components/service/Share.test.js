/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { axe, toHaveNoViolations } from 'jest-axe';

import React from 'react';
import Share from './Share';
import { ToastContext } from '../../App';
import { act } from 'react-dom/test-utils';
import { getModal } from '../../../../testing/testHelpers';
import { render } from 'react-dom';

expect.extend(toHaveNoViolations);

it('renders share a service modal', async () => {
  const container = document.createElement('div');

  await act(async () => {
    render(
      <ToastContext.Provider
        value={{ toastMessage: null, setToastMessage: () => {} }}
      >
        <Share
          onCancel={() => {}}
          serviceName="serviceName"
          serviceUrl="serviceUrl"
        />
      </ToastContext.Provider>,
      container
    );
  });

  expect(getModal()).toMatchSnapshot();
  expect(await axe(getModal())).toHaveNoViolations();
});
