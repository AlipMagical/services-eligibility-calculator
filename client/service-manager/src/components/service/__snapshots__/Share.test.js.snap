// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders share a service modal 1`] = `
<div
  aria-modal="true"
  class="fade action-modal modal show"
  role="dialog"
  style="display: block;"
  tabindex="-1"
>
  <div
    class="modal-dialog"
    role="document"
  >
    <div
      class="modal-content"
    >
      <div
        class="header modal-header"
      >
        <h2>
          Share service
        </h2>
      </div>
      <div
        class="modal-body"
      >
        <p>
          They will be able to edit, publish, and delete this service
        </p>
        <form
          id="share-form"
        >
          <div
            class="form-body"
          >
            <div
              class="row"
            >
              <div
                class="col-sm-3"
              >
                <label
                  class="col-form-label"
                  for="share_email"
                >
                  Email:
                </label>
              </div>
              <div
                class="col"
              >
                <input
                  class="form-control"
                  id="share_email"
                  name="email"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="float-right"
          >
            <button
              class="btn btn-blue-text"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-primary"
            >
              Share
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
